Linked list project - Pull Request

DvdHalliday opened the pull request:
This is a singly linked list with some basic features we all implemented. There is even room for iteration and operations using lambda functions, also there should be no memory leaks on this.

![SLList](https://user-images.githubusercontent.com/109089386/190904739-0511ba84-301a-4fe8-8644-699516b8630c.png)

DvdHalliday committed:
Part 1

I created some basic constructors and that's about it.

DvdHalliday committed:
Part 2

DvdHalliday committed:
Part 3

DvdHalliday committed:
Part 4

DvdHalliday committed:
Part 5 + complements

I finally debugged the whole data structure plus added working versions of  this part

DvdHalliday committed:
Second Part of the Project

Added every feature expected of this part.
Fixed some memory leaks, corrected the file format, fixed some behaviours and improved some others.

DvdHalliday committed:
.sln and .vcxproj files commit

DvdHalliday committed:
Revert "Modifying constructors"

This reverts commit 33c3d42458fc64a59122d97c1929f3628ff71983.

DvdHalliday committed:
Merge branch 'main' into LinkedLists

simonkytr commented on c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h:
```diff
@@ -0,0 +1,315 @@
+		if (IsEmpty())
+		{
+			AddHead(Value);
+			return;
```
why do you return in a void method?

DvdHalliday replied:
To stop the function from running the rest of its instructions.

simonkytr commented on c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h:
```diff
@@ -0,0 +1,315 @@
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
```
wow! you create a function to get a previous node!!

simonkytr commented on c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h:
```diff
@@ -0,0 +1,315 @@
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		int OriginalSize = Size;
```
why do you create a variable to size? you could use just the size variable

DvdHalliday replied:
No I could not, since I'm doing this with the Remove method which will decrease size on the run, this will cause an out of bounds scenario pretty soon.

simonkytr commented on c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h:
```diff
@@ -0,0 +1,315 @@
+		int OriginalSize = Size;
+		for (int i = 0; i < OriginalSize; i++)
+		{
+			Remove(0);
```
do you are deleting just the first element?

DvdHalliday replied:
No, I am deleting every element, it's just that they all become the head once its predecessor gets deleted, I also think this is the most efficient way of doing it since removing the head is the first possible instruction inside the Remove method.

simonkytr commented on c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h:
```diff
@@ -0,0 +1,315 @@
+	void ForEach(const Pred& Predicate)
+	{
+		for (T& Item : *this)
+		{
```
Wow, I didn't have any idea that I could do something like this, nice

DvdHalliday replied:
like *this v;;

DvdHalliday committed:
Doubly LL + Stack + Queue Part

Adding all em files in there

DvdHalliday committed:
Project files update

DvdHalliday committed:
Final adjustments before review

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+		FNode* Next;
+		FNode* Previous;
+
+		FNode(const T& NewData, FNode* NewNext = nullptr, FNode* NewPrevious = nullptr) : Data(NewData), Next(NewNext), Previous(NewPrevious) {}
```
Great usage of the parameters 👍 the FNode* NewNext = nullptr  sets a default value for the parameter if this is not given, right?

DvdHalliday replied:
Yup

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	FNode& GetNodeAt(const int Index) const
+	{
+		if (Index < Size / 2)
+		{
+			FNode* CurrentNode = Head;
+			for (int i = 0; i < Index; i++)
+			{
+				CurrentNode = CurrentNode->Next;
+			}
+
+			return *CurrentNode;
+		}
+
+		FNode* CurrentNode = Tail;
```
Better declaring the `CurrentNode` before the if statement, and then initializing by the value of your need, your way is good, but may be with this recommendation it will be more readable.

DvdHalliday replied:
Yee that does sound sensible, thanks!

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+
+		return *this;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T& operator[](const int Index)
+	{
+		return GetNodeAt(Index).Data;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	const T& operator[](const int Index) const
+	{
+		return GetNodeAt(Index).Data;
+	}
```
this works?? it will not be 
```
GetNode(Index)->Data;
```
??

Sergio6040 replied:
:o okok so GetNode returns a reference, not the pointer, so it is already destructured. nice
it works but isn't consistent :/

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h:
```diff
@@ -0,0 +1,314 @@
+		return *this;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T& operator[](const int Index)
+	{
+		FNode* CurrentNode = Head;
+		for (int i = 0; i < Index; i++)
+		{
+			CurrentNode = CurrentNode->Next;
+		}
+
+		return CurrentNode;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	const T& operator[](const int Index) const
+	{
+		FNode* CurrentNode = Head;
+		for (int i = 0; i < Index; i++)
+		{
+			CurrentNode = CurrentNode->Next;
+		}
+
+		return CurrentNode->Value;
+	}
```
this is not consistent, also here you can use the getNodeAt function.

DvdHalliday replied:
o; you checking on my old stuff huh

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+			return;
+		}
+
+		Tail->Next = new FNode(Data, nullptr, Tail);
```
This also happened to me xd at the FNode constructor is better to put (Data, Previous, Next), it makes your life easier

DvdHalliday replied:
Yee, it kinda does feel far more natural to set the previous first

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* CurrentPtr = PreviousNode.Next;
+			PreviousNode.Next = new FNode(Data, CurrentPtr, &PreviousNode);
+
+			if (Index == Size)
+			{
+				Tail = Tail->Next;
+			}
+			else
+			{
+				CurrentPtr->Previous = PreviousNode.Next;
+			}
```
here you can use GetNodeAt(Index); there is no more need of the Index - 1, it will change some logic but is more clear

DvdHalliday replied:
The thing is, I'm trying to also group up the case in which Index can actually be the same as Size, so I can't in fact call GetNodeAt on such index, which is the main reason i act on the previous one for this is guaranteed to be an actual node

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* CurrentPtr = PreviousNode.Next;
+			PreviousNode.Next = new FNode(Data, CurrentPtr, &PreviousNode);
+
+			if (Index == Size)
+			{
+				Tail = Tail->Next;
+			}
+			else
+			{
+				CurrentPtr->Previous = PreviousNode.Next;
```
also, you keep using .Next What would JD think about you if he see this :C  xd

Sergio6040 replied:
actually I think it's a good way to work

DvdHalliday replied:
Soo... do you approve? or would you still recommend I change it? it's kinda unclear haha

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		for (T& Item : *this)
+		{
+			Predicate(Item);
+		}
```
this is nice, I think I'm gonna borrow it `ctrl + c` `ctrl + v` xd

DvdHalliday replied:
MORE LIKE STEALIN', HUH

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+			if (Predicate(CurrentNode->Value))
+			{
+				CurrentNode = CurrentNode->Next;
+				Remove(i--);
```
:O Remove(i--); my god, BIG BRAIN 😵 this is really nice, good job

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h:
```diff
@@ -0,0 +1,323 @@
+
+		Size--;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		while (Head)
+		{
+			Remove(0);
+		}
+	}
```
this is also very clever  👏

DvdHalliday replied:
Thank you! I actually took this one from a loose comment of JD's during the class

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/Queue.h:
```diff
@@ -0,0 +1,35 @@
+	TSLList<T> List;
+
+public:
+	const T& Front() const
+	{
+		return List.GetHead();
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T& Front()
+	{
+		return List.GetHead();
+	}
```
Great!

Sergio6040 commented on c++ training/LinkedLists/LinkedLists/Stack.h:
```diff
@@ -0,0 +1,35 @@
+#pragma once
+#include "SinglyLinkedList.h"
+
+template<typename T>
+class TStack
+{
+private:
+	TSLList<T> List;
+
+public:
+	const T& Top() const
+	{
+		return List.GetHead();
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T& Top()
+	{
+		return List.GetHead();
+	}
```
Everything good, everything correct, i'm so glad. Suena mejor en español xd

DvdHalliday replied:
perhaps everything is correct or everything seems correct haha!

Sergio6040 reviewed (approved):
great work 👍

DvdHalliday committed:
Addressing Sergio's comments

I did change mostly everything except the GetNodeAt behaviour, for it was not quite clear if he agreed or not with how things are done in the end.
